dropshot/src/error.rs

```diff
@@ -316,6 +337,649 @@ impl Error for HttpError {
     }
 }
 
+/// An HTTP 4xx (client error) or 5xx (server error) status code.
+///
+/// This is a refinement of the [`http::StatusCode`] type that is limited to the
+/// error status code ranges. It may be constructed from any
+/// [`http::StatusCode`] using the `TryFrom` implementation, which fails if the
+/// status is not a 4xx or 5xx status code.
+///
+/// Alternatively, constants are provided for known error status codes,
+/// such as [`ErrorStatusCode::BAD_REQUEST`], [`ErrorStatusCode::NOT_FOUND`],
+/// [`ErrorStatusCode::INTERNAL_SERVER_ERROR`], and so on, including those in
+/// the IANA HTTP Status Code Registry](
+/// https://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml).
+/// Using these constants avoids the fallible conversion from an [`http::StatusCode`].
+///
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ErrorStatusCode(http::StatusCode);
+
+macro_rules! error_status_code_constants {
+    ( $($(#[$docs:meta])* $name:ident;)+ ) => {
+        $(
+            $(#[$docs])*
+            pub const $name: Self = Self(http::StatusCode::$name);
+        )+
+    }
+}
+
+macro_rules! impl_status_code_wrapper {
+    (impl StatusCode for $T:ident {
```

Comment:
This is just me being cutesy about the input syntax to the macro; it could just be `$T:ident,  $NotAnError:ident,  $Invalid:ident` instead, but I thought it was fun to make it look like an actual `impl` block. It does also give us a nice way to see which type is which when reading the code. But, if this is confusing, I could get rid of it.